File: sdk/src/test/java/com/cxense/cxensesdk/db/DatabaseHelperTest.java
package com.cxense.cxensesdk.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.modules.junit4.PowerMockRunner;

import java.util.Arrays;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyDouble;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.intThat;
import static org.mockito.ArgumentMatchers.nullable;
import static org.mockito.ArgumentMatchers.startsWith;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.powermock.api.mockito.PowerMockito.doNothing;
import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.api.mockito.PowerMockito.spy;
import static org.powermock.api.mockito.PowerMockito.when;

/**
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2017-08-04).
 */
@RunWith(PowerMockRunner.class)
public class DatabaseHelperTest {
    private DatabaseHelper databaseHelper;
    private SQLiteDatabase db;
    private ContentValues values;
    private DatabaseObject databaseObject;
    private Cursor cursor;

    @Before
    public void setUp() throws Exception {
        Context context = mock(Context.class);
        db = mock(SQLiteDatabase.class);
        values = mock(ContentValues.class);
        databaseObject = mock(DatabaseObject.class);
        cursor = mock(Cursor.class);
        databaseHelper = spy(new DatabaseHelper(context));
        doReturn(db).when(databaseHelper).getWritableDatabase();
        doReturn(db).when(databaseHelper).getReadableDatabase();
        doReturn(values).when(databaseObject).toContentValues();
        doReturn(false).when(cursor).moveToNext();
        doNothing().when(values).putNull(anyString());
        doReturn(1L).when(db).insert(any(), any(), any(ContentValues.class));
        doReturn(1).when(db).update(any(), any(ContentValues.class), any(), any(String[].class));
        doReturn(1).when(db).delete(any(), any(), any());
        doReturn(cursor).when(db).query(any(), any(), any(), any(), any(), any(), any(), any());
    }

    @Test
    public void parseCursor() throws Exception {
        int[] types = new int[]{
                Cursor.FIELD_TYPE_NULL, Cursor.FIELD_TYPE_INTEGER, Cursor.FIELD_TYPE_FLOAT,
                Cursor.FIELD_TYPE_STRING, Cursor.FIELD_TYPE_BLOB
        };
        String[] columns = Arrays.stream(types).mapToObj(String::valueOf).toArray(String[]::new);
        when(cursor.getColumnNames()).thenReturn(columns);
        when(cursor.getType(anyInt())).thenReturn(Cursor.FIELD_TYPE_NULL);
        when(cursor.getType(anyInt())).thenAnswer(invocation -> {
            int arg = invocation.getArgument(0);
            if (arg < types.length)
                return types[arg];
            return types[0];
        });
        DatabaseHelper.parseCursor(cursor, values);
        verify(values).putNull(anyString());
        verify(values).put(anyString(), anyLong());
        verify(values).put(anyString(), anyDouble());
        verify(values).put(anyString(), nullable(String.class));
        verify(values).put(anyString(), nullable(byte[].class));
    }

    @Test
    public void cursorRowToContentValues() throws Exception {
        when(cursor.getColumnNames()).thenReturn(new String[]{});
        DatabaseHelper.cursorRowToContentValues(cursor);
    }

    @Test
    public void onCreate() throws Exception {
        doNothing().when(databaseHelper).migrate(any(), anyInt());
        databaseHelper.onCreate(db);
        verify(databaseHelper, times(DatabaseHelper.DATABASE_VERSION)).migrate(eq(db), anyInt());
    }

    @Test
    public void onUpgrade() throws Exception {
        int oldVersion = 3, newVersion = 5;
        doNothing().when(databaseHelper).migrate(any(), anyInt());
        databaseHelper.onUpgrade(db, oldVersion, newVersion);
        verify(databaseHelper, times(newVersion - oldVersion))
                .migrate(eq(db), intThat(argument -> argument >= oldVersion && argument < newVersion));
    }

    @Test
    public void save() throws Exception {
        databaseObject.id = 1;
        assertEquals(1, databaseHelper.save(databaseObject));
    }

    @Test
    public void saveNewObject() throws Exception {
        databaseObject.id = null;
        assertEquals(1L, databaseHelper.save(databaseObject));
    }

    @Test
    public void delete() throws Exception {
        databaseObject.id = 1;
        assertEquals(1, databaseHelper.delete(databaseObject));
    }

    @Test
    public void deleteNotExistObject() throws Exception {
        databaseObject.id = null;
        assertEquals(-1, databaseHelper.delete(databaseObject));
    }

    @Test
    public void deleteFullArgs() throws Exception {
        assertEquals(1, databaseHelper.delete("table", null, null));
    }

    @Test
    public void query() throws Exception {
        assertNotNull(databaseHelper.query("table", new String[]{}, null,
                null, null, null, null));
    }

    @Test
    public void queryFullArgs() throws Exception {
        assertNotNull(databaseHelper.query("table", new String[]{}, null,
                null, null, null, null, null));
    }

    @Test
    public void migrate() throws Exception {
        for (int i = 0; i < DatabaseHelper.DATABASE_VERSION; i++) {
            databaseHelper.migrate(db, i);
            verify(db).execSQL(startsWith("CREATE TABLE " + EventRecord.TABLE_NAME));
        }
    }

}

File: sdk/src/test/java/io/piano/android/cxense/db/DatabaseHelperTest.kt
package io.piano.android.cxense.db

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.provider.BaseColumns
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.anyOrNull
import com.nhaarman.mockitokotlin2.doNothing
import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.spy
import com.nhaarman.mockitokotlin2.stub
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import org.mockito.ArgumentMatchers.intThat
import org.mockito.ArgumentMatchers.startsWith
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

class DatabaseHelperTest {
    private lateinit var databaseHelper: DatabaseHelper
    private lateinit var eventRecord: EventRecord

    private val context: Context = mock()
    private val cursor: Cursor = mock {
        on { moveToNext() } doReturn false
    }
    private val db: SQLiteDatabase = mock {
        on {
            query(any(), any(), anyOrNull(), anyOrNull(), anyOrNull(), anyOrNull(), any(), anyOrNull())
        } doReturn cursor
    }
    private val contentValues: ContentValues = mock()

    @BeforeTest
    fun setUp() {
        databaseHelper = spy(DatabaseHelper(context))
        doReturn(db).`when`(databaseHelper).writableDatabase
        doReturn(db).`when`(databaseHelper).readableDatabase

        eventRecord = mock {
            on { toContentValues() } doReturn contentValues
        }
    }

    @Test
    fun onCreate() {
        doNothing().`when`(databaseHelper).migrate(any(), any())
        databaseHelper.onCreate(db)
        verify(databaseHelper, times(DatabaseHelper.DATABASE_VERSION)).migrate(eq(db), any())
    }

    @Test
    fun onUpgrade() {
        val oldVersion = 3
        val newVersion = 5
        doNothing().`when`(databaseHelper).migrate(any(), any())
        databaseHelper.onUpgrade(db, oldVersion, newVersion)
        verify(databaseHelper, times(newVersion - oldVersion))
            .migrate(
                eq(db),
                intThat { it in oldVersion until newVersion }
            )
    }

    @Test
    fun migrate() {
        databaseHelper.migrate(db, 0)
        verify(db).execSQL(startsWith("CREATE TABLE "))
    }

    @Test
    fun save() {
        databaseHelper.save(eventRecord)
        verify(eventRecord).toContentValues()
        verify(db).update(any(), eq(contentValues), eq("${BaseColumns._ID} = ?"), any())
    }

    @Test
    fun saveNewRecord() {
        eventRecord.stub {
            on { id } doReturn null
        }
        databaseHelper.save(eventRecord)
        verify(eventRecord).toContentValues()
        verify(db).insert(any(), eq(null), eq(contentValues))
    }

    @Test
    fun deleteRecord() {
        doReturn(1).`when`(databaseHelper).delete(any(), any())
        assertEquals(1, databaseHelper.delete(eventRecord))
        verify(databaseHelper).delete(eq("${BaseColumns._ID} = ?"), any())
    }

    @Test
    fun deleteNotExistRecord() {
        eventRecord.stub {
            on { id } doReturn null
        }
        assertEquals(-1, databaseHelper.delete(eventRecord))
    }

    @Test
    fun deleteFullArgs() {
        assertEquals(0, databaseHelper.delete(null))
        verify(db).delete(any(), eq(null), eq(emptyArray()))
    }

    @Test
    fun query() {
        databaseHelper.query()
        verify(db).query(any(), any(), anyOrNull(), anyOrNull(), anyOrNull(), anyOrNull(), anyOrNull(), anyOrNull())
    }
}
